File: src/ur-fountain.c
/* BC-UR fountain fragment selection. See ur-fountain.h. */

#include "ur-fountain.h"

#include <string.h>

#include "sha2.h"

static uint64_t rotl(uint64_t x, int k)
{
    return (x << k) | (x >> (64 - k));
}

void ur_rng_seed(UrRng *rng, const uint8_t digest[32])
{
    for (int i = 0; i < 4; i++) {
        uint64_t v = 0;
        for (int n = 0; n < 8; n++) {
            v <<= 8;
            v |= digest[i * 8 + n];
        }
        rng->s[i] = v;
    }
}

uint64_t ur_rng_next(UrRng *rng)
{
    uint64_t *s = rng->s;
    const uint64_t result = rotl(s[1] * 5, 7) * 9;
    const uint64_t t = s[1] << 17;

    s[2] ^= s[0];
    s[3] ^= s[1];
    s[1] ^= s[2];
    s[0] ^= s[3];
    s[2] ^= t;
    s[3] = rotl(s[3], 45);

    return result;
}

double ur_rng_next_double(UrRng *rng)
{
    /* 2^64 exactly, written so the compiler cannot be tempted to fold this
       differently from the reference. */
    const double m = 18446744073709551616.0;
    return (double)ur_rng_next(rng) / m;
}

uint64_t ur_rng_next_int(UrRng *rng, uint64_t low, uint64_t high)
{
    return (uint64_t)(ur_rng_next_double(rng) * (double)(high - low + 1)) + low;
}

/*
 * Walker's alias method over the degree distribution 1/1, 1/2 ... 1/seq_len.
 *
 * Built exactly as bc-ur builds it, including the reversed index order it
 * notes as a variance from Schwarz, because the table it produces is part of
 * the wire behaviour: a different but equally valid alias table would pick
 * different fragments and the two ends would stop agreeing.
 */
static size_t choose_degree(uint32_t seq_len, UrRng *rng,
                            UrFountainScratch *sc)
{
    const int n = (int)seq_len;

    double sum = 0.0;
    for (int i = 1; i <= n; i++) {
        sum += 1.0 / (double)i;
    }

    /* P[i] = p_i * n / sum, the scaled probabilities the method works over. */
    double *P = sc->probs;
    for (int i = 0; i < n; i++) {
        P[i] = (1.0 / (double)(i + 1)) * (double)n / sum;
    }

    int16_t *small = sc->remaining;       /* reused as the "small" stack */
    int16_t *large = sc->aliases;         /* and this as the "large" stack */
    int n_small = 0, n_large = 0;

    for (int i = n - 1; i >= 0; i--) {
        if (P[i] < 1.0) {
            small[n_small++] = (int16_t)i;
        } else {
            large[n_large++] = (int16_t)i;
        }
    }

    double probs[UR_MAX_PARTS];
    int16_t aliases[UR_MAX_PARTS];
    for (int i = 0; i < n; i++) {
        probs[i] = 0.0;
        aliases[i] = 0;
    }

    while (n_small > 0 && n_large > 0) {
        const int a = small[--n_small];
        const int g = large[--n_large];
        probs[a] = P[a];
        aliases[a] = (int16_t)g;
        P[g] += P[a] - 1.0;
        if (P[g] < 1.0) {
            small[n_small++] = (int16_t)g;
        } else {
            large[n_large++] = (int16_t)g;
        }
    }
    while (n_large > 0) {
        probs[large[--n_large]] = 1.0;
    }
    while (n_small > 0) {
        /* Reachable only through numeric instability, per the reference. */
        probs[small[--n_small]] = 1.0;
    }

    const double r1 = ur_rng_next_double(rng);
    const double r2 = ur_rng_next_double(rng);
    const int i = (int)((double)n * r1);
    const int chosen = (r2 < probs[i]) ? i : aliases[i];

    return (size_t)chosen + 1;
}

bool ur_fountain_fragments(uint32_t seq_num, uint32_t seq_len, uint32_t checksum,
                           uint8_t mask[UR_PART_MASK_BYTES],
                           UrFountainScratch *scratch)
{
    if (mask == NULL || scratch == NULL) {
        return false;
    }
    if (seq_num == 0 || seq_len == 0 || seq_len > UR_MAX_PARTS) {
        return false;
    }

    /* The first seq_len parts are the plain fragments. Generating only those
       is a complete, non-fountain encoding, which is what a sender with a big
       screen and a cooperative receiver can get away with. */
    if (seq_num <= seq_len) {
        memset(mask, 0, UR_PART_MASK_BYTES);
        const uint32_t idx = seq_num - 1;
        mask[idx / 8] |= (uint8_t)(1u << (idx % 8));
        return true;
    }

    const uint8_t seed_in[8] = {
        (uint8_t)(seq_num >> 24), (uint8_t)(seq_num >> 16),
        (uint8_t)(seq_num >> 8),  (uint8_t)seq_num,
        (uint8_t)(checksum >> 24), (uint8_t)(checksum >> 16),
        (uint8_t)(checksum >> 8),  (uint8_t)checksum,
    };
    uint8_t digest[32];
    sha256_Raw(seed_in, sizeof seed_in, digest);

    UrRng rng;
    ur_rng_seed(&rng, digest);

    const size_t degree = choose_degree(seq_len, &rng, scratch);

    /* Fisher-Yates by removal, exactly as the reference shuffles: draw an
       index from what is left, take it, close the gap. Only the first `degree`
       draws matter, but the draws themselves must match. */
    int16_t *remaining = scratch->remaining;
    for (uint32_t i = 0; i < seq_len; i++) {
        remaining[i] = (int16_t)i;
    }

    memset(mask, 0, UR_PART_MASK_BYTES);
    size_t left = seq_len;
    for (size_t taken = 0; taken < degree && left > 0; taken++) {
        const uint64_t index = ur_rng_next_int(&rng, 0, (uint64_t)left - 1);
        const int16_t item = remaining[index];
        for (size_t j = (size_t)index; j + 1 < left; j++) {
            remaining[j] = remaining[j + 1];
        }
        left--;
        mask[item / 8] |= (uint8_t)(1u << (item % 8));
    }

    return true;
}

uint32_t ur_mask_count(const uint8_t mask[UR_PART_MASK_BYTES], uint32_t seq_len)
{
    uint32_t n = 0;
    for (uint32_t i = 0; i < seq_len && i < UR_MAX_PARTS; i++) {
        if (mask[i / 8] & (1u << (i % 8))) {
            n++;
        }
    }
    return n;
}

File: app/packages/core/src/ur-fountain.ts
/**
 * The BC-UR fountain code: which fragments a given part carries.
 *
 * Mirror of `src/ur-fountain.c`. Held to it by `test/ur-fountain-vectors.json`,
 * which the firmware emits.
 *
 * ---------------------------------------------------------------------------
 * Why this one is the risky mirror
 *
 * The subset a mixed part carries is never transmitted. Both ends derive it
 * from the part number, the fragment count and the message checksum, so an
 * implementation that derives it differently does not fail loudly — it
 * silently XORs the wrong fragments together and the assembled message fails
 * its checksum, which looks exactly like a camera misread.
 *
 * Worse, the derivation runs through IEEE-754 doubles. Two implementations
 * agreeing here is also a claim that C and JavaScript agree bit for bit about
 * double arithmetic, which is true but is not something to assume. That is why
 * the vector file is large and weighted to the mixed path rather than being a
 * handful of spot checks.
 *
 * ---------------------------------------------------------------------------
 * 64-bit arithmetic
 *
 * Xoshiro256** is defined over uint64. JavaScript numbers cannot represent it,
 * so this uses BigInt throughout and converts to a double only where the
 * reference does. Slower than it would be with two 32-bit halves, and not on
 * any hot path: a part is chosen once per scanned frame.
 */

import { sha256 } from "@noble/hashes/sha256";

/** Largest fragment count we will handle; matches UR_MAX_PARTS in the firmware. */
export const UR_MAX_PARTS = 128;

const MASK64 = (1n << 64n) - 1n;
const TWO_POW_64 = 18446744073709551616; // 2^64 as a double, exactly

function rotl(x: bigint, k: bigint): bigint {
  return ((x << k) | (x >> (64n - k))) & MASK64;
}

export class Xoshiro256 {
  private s: [bigint, bigint, bigint, bigint];

  /** Seed from a 32-byte digest, big-endian per 64-bit word. */
  constructor(digest: Uint8Array) {
    if (digest.length !== 32) {
      throw new Error("xoshiro256: seed must be 32 bytes");
    }
    const w = (o: number): bigint => {
      let v = 0n;
      for (let i = 0; i < 8; i++) v = (v << 8n) | BigInt(digest[o + i]!);
      return v;
    };
    this.s = [w(0), w(8), w(16), w(24)];
  }

  next(): bigint {
    const s = this.s;
    const result = (rotl((s[1] * 5n) & MASK64, 7n) * 9n) & MASK64;
    const t = (s[1] << 17n) & MASK64;

    s[2] ^= s[0];
    s[3] ^= s[1];
    s[1] ^= s[2];
    s[0] ^= s[3];
    s[2] ^= t;
    s[3] = rotl(s[3], 45n);

    return result;
  }

  nextDouble(): number {
    return Number(this.next()) / TWO_POW_64;
  }

  nextInt(low: number, high: number): number {
    return Math.floor(this.nextDouble() * (high - low + 1)) + low;
  }
}

/**
 * Walker's alias method over the degree distribution 1/1, 1/2 ... 1/seqLen.
 *
 * Built exactly as bc-ur builds it, including the reversed index order it flags
 * as a variance from Schwarz. That ordering is part of the wire behaviour: a
 * different but equally valid alias table picks different fragments, and the
 * two ends stop agreeing without either of them being wrong on its own terms.
 */
function chooseDegree(seqLen: number, rng: Xoshiro256): number {
  const n = seqLen;

  let sum = 0;
  for (let i = 1; i <= n; i++) sum += 1 / i;

  const P = new Float64Array(n);
  for (let i = 0; i < n; i++) P[i] = ((1 / (i + 1)) * n) / sum;

  const small: number[] = [];
  const large: number[] = [];
  for (let i = n - 1; i >= 0; i--) {
    (P[i]! < 1 ? small : large).push(i);
  }

  const probs = new Float64Array(n);
  const aliases = new Int32Array(n);

  while (small.length > 0 && large.length > 0) {
    const a = small.pop()!;
    const g = large.pop()!;
    probs[a] = P[a]!;
    aliases[a] = g;
    P[g] = P[g]! + P[a]! - 1;
    (P[g]! < 1 ? small : large).push(g);
  }
  while (large.length > 0) probs[large.pop()!] = 1;
  /* Reachable only through numeric instability, per the reference. */
  while (small.length > 0) probs[small.pop()!] = 1;

  const r1 = rng.nextDouble();
  const r2 = rng.nextDouble();
  const i = Math.floor(n * r1);
  return (r2 < probs[i]! ? i : aliases[i]!) + 1;
}

/**
 * The fragments part `seqNum` carries. `seqNum` is 1-based, as on the wire.
 *
 * Parts 1..seqLen select exactly one fragment each, so a sender that emits only
 * those has sent a complete message and a receiver never has to XOR anything.
 */
export function chooseFragments(
  seqNum: number,
  seqLen: number,
  checksum: number,
): number[] {
  if (seqNum < 1 || seqLen < 1 || seqLen > UR_MAX_PARTS) {
    throw new Error(`fountain: seqNum ${seqNum} of seqLen ${seqLen} is out of range`);
  }
  if (seqNum <= seqLen) {
    return [seqNum - 1];
  }

  const seed = new Uint8Array(8);
  const dv = new DataView(seed.buffer);
  dv.setUint32(0, seqNum >>> 0, false);
  dv.setUint32(4, checksum >>> 0, false);

  const rng = new Xoshiro256(sha256(seed));
  const degree = chooseDegree(seqLen, rng);

  /* Fisher-Yates by removal, exactly as the reference shuffles: draw an index
     from what is left, take it, close the gap. Only the first `degree` draws
     are used, but the draws themselves have to match. */
  const remaining: number[] = [];
  for (let i = 0; i < seqLen; i++) remaining.push(i);

  const chosen: number[] = [];
  for (let taken = 0; taken < degree && remaining.length > 0; taken++) {
    const index = rng.nextInt(0, remaining.length - 1);
    chosen.push(remaining[index]!);
    remaining.splice(index, 1);
  }
  return chosen.sort((a, b) => a - b);
}
